### PR TITLE
Add SignalR Action

### PR DESCRIPTION
**Pull request by julienM77**
Add SignalR Azure push message

Just a litlle validation error to convert Action to Enum (thanks for your help)

It use package https://www.nuget.org/packages/Microsoft.Azure.SignalR.Management

image

### CODE REVIEW

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        IconImage = "<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 32 32'><path d='M.011 16L0 6.248l12-1.63V16zM14 4.328L29.996 2v14H14zM30 18l-.004 14L14 29.75V18zM12 29.495L.01 27.851.009 18H12z'/></svg>",
+        IconColor = "#1566BF",
+        Display = "Send to Azure SignalR",
+        Description = "Send an message to azure SignalR.",
+        ReadMore = "https://azure.microsoft.com/fr-fr/services/signalr-service/")]
+    public sealed record SignalRAction : RuleAction
+    {
+        [LocalizedRequired]
+        [Display(Name = "Connection", Description = "The connection string to the Signal R Azure.")]
+        [DataType(DataType.Text)]
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs:21]: DataTypeAttribute is not supported anymore. There is a new one: https://github.com/Squidex/squidex/blob/master/backend/extensions/Squidex.Extensions/Actions/AzureQueue/AzureQueueAction.cs#L28

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
+﻿using System.Collections.Generic;
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs:1]: Please check if there are warnings from roslyn, e.g. header is missing and I am not sure if all usings are needed.

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            if (ActionType == ActionTypeEnum.GROUPS && (string.IsNullOrWhiteSpace(Group) || Group.IndexOf("\n", System.StringComparison.OrdinalIgnoreCase) < 0))
+            {
+                yield return new ValidationError("Group must be specified and multiple with 'GROUPS' Action.", nameof(HubName));
+            }
+        }
+    }
+
+    public enum ActionTypeEnum
+    {
+        BROADCAST,
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs:86]: Why are these uppercase?

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+using Microsoft.Azure.SignalR.Management;
+using Squidex.Domain.Apps.Core.HandleRules;
+using Squidex.Domain.Apps.Core.Rules.EnrichedEvents;
+
+namespace Squidex.Extensions.Actions.SignalR
+{
+    public sealed class SignalRActionHandler : RuleActionHandler<SignalRAction, SignalRJob>
+    {
+        private readonly ClientPool<(string ConnectionString, string HubName), IServiceManager> clients;
+        public SignalRActionHandler(RuleEventFormatter formatter)
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs:13]: Please add newline after fields.

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+            return Result.Complete();
+        }
+    }
+
+    public sealed class SignalRJob
+    {
+        public string ConnectionString { get; set; }
+
+        public string HubName { get; set; }
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs:97]: Please add newline after each property.

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            switch (job.Action)
+            {
+                case ActionTypeEnum.BROADCAST:
+                    await signalRContext.Clients.All.SendAsync(methodeName, job.Payload);
+                    break;
+                case ActionTypeEnum.USER:
+                    await signalRContext.Clients.User(job.User).SendAsync(methodeName, job.Payload);
+                    break;
+                case ActionTypeEnum.USERS:
+                    var userIds = job.User.Split('\n');
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs:75]: This should be done in the CreateJobAsync

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRPlugin.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+﻿using Microsoft.Extensions.Configuration;
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRPlugin.cs:1]: Header...

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+                case ActionTypeEnum.GROUP:
+                    await signalRContext.Clients.Group(job.Group).SendAsync(methodeName, job.Payload);
+                    break;
+                case ActionTypeEnum.GROUPS:
+                    var groupIds = job.Group.Split('\n');
+                    await signalRContext.Clients.Groups(groupIds).SendAsync(methodeName, job.Payload);
+                    break;
+            }
+
+            await signalRContext.DisposeAsync();
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs:87]: Please use ´await using (...)` here

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        public string ConnectionString { get; set; }
+
+        [LocalizedRequired]
+        [Display(Name = "Hub Name", Description = "The name of the hub.")]
+        [DataType(DataType.Text)]
+        [Formattable]
+        public string HubName { get; set; }
+
+        [LocalizedRequired]
+        [Display(Name = "Action", Description = "BROADCAST -> send to all User, USER => send to one user, USERS => send to list of users, GROUP => send to group, GROUPS => send to list of groups.")]
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs:32]: You are using different arrows here. I am not sure, but I think you can also use markdown here.

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            {
+                yield return new ValidationError("Group must be specified and multiple with 'GROUPS' Action.", nameof(HubName));
+            }
+        }
+    }
+
+    public enum ActionTypeEnum
+    {
+        BROADCAST,
+        USER,
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs:87]: Can you not make User/Users and Group/Groups dynamic? Depending how many users have been defined?

**Review by SebastianStehle — CHANGES_REQUESTED** [review]
I have a few change requests.

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs (hunk 1 above)
Comment (julienM77) [reply backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs:21]: Ok, I didn't have my upstream up to date with this evolution, it's corrected

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs (hunk 2 above)
Comment (julienM77) [reply backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs:1]: it's corrected for header, all using are useful

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs (hunk 4 above)
Comment (julienM77) [reply backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs:13]: Indeed, it was just a test, I'm switching back to CamelCase

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs (hunk 5 above)
Comment (julienM77) [reply backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs:97]: newline added

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs (hunk 6 above)
Comment (julienM77) [reply backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs:75]: move to CreateJobAsync

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRPlugin.cs (hunk 7 above)
Comment (julienM77) [reply backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRPlugin.cs:1]: it's corrected for header

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs (hunk 8 above)
Comment (julienM77) [reply backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs:87]: ok, delete DisposeAsync and use 'await using(...)'

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs (hunk 10 above)
Comment (julienM77) [reply backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs:87]: I change the code and action for that

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs (hunk 9 above)
Comment (julienM77) [reply backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRAction.cs:32]: I change the first arrows, markdown doesn't seem to work
image

**Comment by julienM77** [comment]
New commit for the fixes.
For information I found out why the validation was crashing, when saving, for the conversion of ActionType to Enum, I think ActionType is already reserved by your use (name of the action, here SignalR) and so my property was problematic. I renamed it to just Action and it's fine.

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+            if (!string.IsNullOrWhiteSpace(action.Payload))
+            {
+                requestBody = await FormatAsync(action.Payload, @event);
+            }
+            else
+            {
+                requestBody = ToEnvelopeJson(@event);
+            }
+
+            string[] users = new string[0];
```
Comment (SebastianStehle) [inline backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs:52]: Don't make it so complicated ;)

```
var users = await FormatAsync(action.User, @event);
var userArray = users.Split('\n');
```

And then in the execute

```
if (job.Users.Count == 1)
{
    await signalRContext.Clients.User(job.User).SendAsync(methodeName, job.Payload);
}
else
{
    // But what happens if you pass in only one user? Do you really need the if?
    await signalRContext.Clients.Users(job.Users).SendAsync(methodeName, job.Payload);
}
```

**Review by SebastianStehle — CHANGES_REQUESTED** [review]
I think it can be further simplified.

File: backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs (hunk 11 above)
Comment (julienM77) [reply backend/extensions/Squidex.Extensions/Actions/SignalR/SignalRActionHandler.cs:52]: In fact I wanted to shield the transition from the object to the Client methods.
User can only take a string as parameter, Users can only take an array of strings. Afterwards, if the Array for Users contains only one item, it's not a problem

**Comment by julienM77** [comment]
Ok, code reduction and I use more than just the Clients and Groups method as an array of strings of an item works very well.
Also reduced the parameterization to have only one target field :)

**Comment by SebastianStehle** [comment]
Thx a lot for your effort and the good results and the patience to implement all change requests.. I have improved the wording a little bit.

**Comment by julienM77** [comment]
Thanks for your explanations and the integration of my code.
More on this in the next PR ;)

**Pull request merged**